src/functional/iterator/transforms/reduction_utils.py

```diff
@@ -0,0 +1,133 @@
+import abc
+import types
+from collections.abc import Iterable, Iterator
+from typing import TypeGuard
+
+from functional import common
+
+
+class _Expr(abc.ABC):
+    ...
+
+
+class _SymRef(_Expr, abc.ABC):
+    @property
+    @abc.abstractmethod
+    def id(self):  # noqa: A003
+        ...
+
+
+class _FunCall(_Expr, abc.ABC):
+    @property
+    @abc.abstractmethod
+    def fun(self):
+        ...
+
+    @property
+    @abc.abstractmethod
+    def args(self):
+        ...
+
+
+class _OffsetLiteral(_Expr, abc.ABC):
+    @property
+    @abc.abstractmethod
+    def value(self):
+        ...
```

Comment:
Unlike in the `Protocol` case, none of these classes needs to inherit directly from `abc.ABC`, since they already inherit indirectly via `_Expr`.

Reply:
Makes sense! In my first approach I didn't reproduce the expr hierarchy, but of course I got some mypy issues, then later I added `_Expr`.